Rewrite cutting docs (#1624)
pace-neutrons/Horace: documentation/user_docs/docs/manual/Cutting_data_of_interest_from_SQW_files_and_objects.rst

###################################################

.. |SQW| replace:: S(**Q**, :math:`\omega{}`)

.. |Q| replace:: :math:`|\textbf{Q}|`

Horace stores the data produced in a neutron experiment in a 4-D dataset
containing the full information about an experiment, which are called the
"pixels". These represent the neutron events measured during an
experiment.

Horace also creates a 4-dimensional histogram (binned) of these "pixels"
represented in reciprocal space (``hkl-dE``) which we call the "image". This
carries only limited information about to the original data, mostly only the
averages of data in the bins.

.. note::

   For the differences between ``dnd`` and ``sqw`` objects see: :ref:`here
   <manual/FAQ:What is the difference between sqw and dnd objects>`

The entire dataset is often too large to fit in the memory of most
computers. Generally, rather than attempting to deal with the full experimental
data, a user works with smaller objects, extracted from the full dataset using
``cut``.

cut
===

``cut`` takes a (or multiple) section(s) of data from an ``sqw`` or ``dnd``
object or file, discards the pixels which lie outside of the binning regions

> (described `below <#binning-arguments>`_) and accumulates them into a histogram

Author: A histogram can be N-dimensional. There is no analogy here or bar-chart.

given.

``cut`` can produce objects of the same or reduced size and dimensions. The
result of a cut is itself an ``sqw`` or ``dnd`` object which can be further
plotted, manipulated, etc.

.. note::

   If the resulting object is estimated by Horace to be too big to fit in
   memory, it will be saved to disc, but will appear mostly indistinguishable
   from a memory-backed ``sqw`` object, supporting most operations. We call this
   a "file-backed ``sqw``" (for more detail see: `File- and memory-backed
   cuts`_). When the manual refers to an ``sqw`` object both "file-backed" and
   "memory-backed" are included.

   .. warning::

      File-backed objects may be significantly slower to operate on than memory
      backed ones.

The inputs are as follows:

.. code-block:: matlab

   my_cut = cut(data_source[, proj], p1_bin, p2_bin, p3_bin, p4_bin[, '-nopix'][, sym][, filename])

where:

- `data_source <#datasource>`_ is either a string giving the full filename (including
  path) of an input sqw file or a variable containing a ``sqw`` object.
- `proj <#projection-proj>`_ defines the axes and origin of the cut including

>   the shape of the region to extract and the representation in the resulting

Author: It means the meaning of the axes and in principle a spherically-oriented plot

``line_proj`` which defines a (usually orthogonal, but not necessarily) system
of linear coordinates from a set of basis vectors.

The complete signature for ``line_proj`` is:

.. code-block:: matlab

   proj = line_proj(u, v[, w][, nonorthogonal][, type][, alatt, angdeg][, offset][, label][, title][, lab1][, lab2][, lab3][, lab4]);

Where:

* ``u``

  3-vector in reciprocal space :math:`(h,k,l)` specifying first viewing axis.

* ``v``

  3-vector in reciprocal space :math:`(h,k,l)` in the plane of the second viewing axis.

* ``w``

  3-vector of in reciprocal space :math:`(h,k,l)` of the third viewing axis.

.. note::

   The first viewing axis is strictly defined to be ``u``

   The second viewing axis is constructed by default to be in the plane of ``u``
   and ``v`` and perpendicular to ``u``.

   The third viewing axes is by default defined as the cross product of the first
   two.

>    The fourth viewing axis is always energy and cannot be specified.

Author: ```suggestion
   The fourth viewing axis is always energy and cannot be modified.
```

    However, it is advised that besides ``u`` and ``v`` arguments are passed as
    key-value pairs.

    Alternatively you may define some parameters initally, and define others
    programmatically:

    .. code-block:: matlab

       proj = line_proj([0,1,0],[0,0,1]);
       proj.type = 'aaa';
       proj.title = 'My linear cut';

    Both forms result in the same object

Non-orthogonal axes
___________________

You may choose to use non-orthogonal axes (c.f. `here <#nonortho>`_), e.g.:

.. code-block:: matlab

   proj = line_proj([1 0 0], [0 1 0], [0 0 1], 'nonorthogonal', true);

The figure below shows the difference between ``nonorthogonal=false`` and
``nonorthogonal=true`` for plotting fake "Bragg reflections" from a
non-orthogonal lattice (``alatt=[2,2,4]``, ``angdeg=[90,90,70]``) where the
reflections occur in the r.l.u. points where ``hkl`` coordinates are integers.

 .. figure:: ../images/orthogonal_vs_nonorthogonal_proj.png
   :align: center
   :width: 800px
   :alt: 2d cuts ortho and non-ortho.

   Sample plot for case where projection is a) orthogonal and b) non-orthogonal.

We can see that for the ``orthogonal=true`` case the reciprocal lattice

> maintains its symmetry, but the ``orthogonal=false`` image produces nice axes but

Author: Nor am I. Will change

a skewed image.

.. note::

   These images are produced by the following code:

   .. code-block:: matlab

      function w = reflection(h, k, l, e, p)
           grid_h = round(h);
           grid_k = round(k);
           grid_l = round(l);
           w = p(1)*exp(-((h-grid_h).^2+(k-grid_k).^2+(l-grid_l).^2)/p(2));
      end

      function plot_cuts()
           proj = line_proj([1, 0, 0], [0, 1, 0], [], false, 'rrr', [2, 2, 4], [90, 90, 70]);
           ax = line_axes('nbins_all_dims', [200, 200, 1, 1], 'img_range', [-4, -3, -0.1, -5;4, 3, 0.1, 5]);
           tsqw = sqw.generate_cube_sqw(ax, proj);
           tsqw = sqw_eval(tsqw, @reflection, [1, 0.01]);
           plot(tsqw)
           keep_figure
           proj = line_proj([1, 0, 0], [0, 1, 0], [], true, 'rrr', [2, 2, 4], [90, 90, 70]);
           tso = sqw.generate_cube_sqw(ax, proj);
           tso = sqw_eval(tso, @reflection, [1, 0.01]);
           plot(tso)
      end

   which plots 2D exponential decay around points where ``h,k,l`` are integers.

>    Note that this is not normal usage of a projection. The construction of

Author: @cmarooney-stfc  Does this address the comment below? 
```suggestion
   Note that this is not normal usage of a projection: the construction of
```

   ``line_proj`` here uses positional arguments and because we are building a

>    fake ``sqw`` object, the projection needs the lattice explicitly defined.

Author: It is in the notes above and this explains how it's not normal.

and :math:`[0,0,1]` directions:

.. code-block:: matlab

    data_source = 'Fe_ei401.sqw';
    proj = line_proj([0, 1, 0], [0, 0, 1], 'type', 'aaa');
    w2 = cut(data_source, proj, [-4.5, 0.1, 14.5], [-5, 0.1, 5], [-0.1, 0.1], [-10, 10]);
    plot(w2);

.. note::

   You do not need to provide a lattice for the projection as ``cut`` will use
   the lattice parameters from the ``sqw`` object.

The code produces:

.. figure:: ../images/iv_hkl.png
   :align: center
   :width: 800px
   :alt: 2d cut

   MAPS Fe data; reciprocal space covered by MAPS for an iron sample
   with incident neutron energy of 401meV.

The cut with the same parameters as above at higher energy transfer

.. code-block:: matlab

   w2 = cut(data_source, proj, [-4.5, 0.1, 14.5], [-5, 0.1, 5], [-0.1, 0.1], [50, 60]);
   plot(w2);

shows clear spin waves:

.. figure:: ../images/iv_hkl_dE.png
   :align: center
   :width: 800px
   :alt: 2d cut instrument view energy transfer.

   MAPS Fe Data; reciprocal space covered by MAPS for an iron sample
   with incident neutron energy of 401meV. Energies integrated between [50, 60].

``line_proj`` 1D cut example
____________________________

> The sample cut along the direction :math:`[1,1,0]` (note the projection's ``u``,

Author: Accidentally missed this bit, will fix

  perpendicular to :math:`e_z`

  Mathematically this is defined as:

.. math::

   \cos\left(\theta{}\right) = \frac{\vec{q}\cdot\vec{e_z}}{\left|q\right|\cdot\left|e_z\right|}

.. note::

   :math:`\theta` is in the range between :math:`0^{\circ}` and :math:`180^{\circ}`

* :math:`\phi`

  is the angle measured between the :math:`e_x`-:math:`e_z` plane to the vector
  (:math:`\vec{q}`), i.e. :math:`0^{\circ}` lies in the :math:`e_x`-:math:`e_z`
  plane and :math:`90^{\circ}` is normal to :math:`e_x`-:math:`e_z` plane
  (i.e. parallel to :math:`e_y`).

  Mathematically this is defined as:

.. math::

   \sin\left(\phi{}\right) = \frac{\vec{q}\cdot\vec{e_y}}{\left|q\right|\cdot\left|e_y\right|}

.. note::

  :math:`\phi` is in the range between :math:`-180^{\circ}` and :math:`180^{\circ}`

* :math:`E`

  is the energy transfer as defined in ``line_proj``

.. figure:: ../images/sphere_proj_graph.jpg
   :align: center
   :width: 400px
   :alt: spherical coordinate system.

   Spherical coordinate system used by ``sphere_proj``.

.. note::

   A spherical projection currently does not have the ability to be

>    rescaled to |Q|.

Reviewer: may be worth expanding exactly what this means

When it comes to cutting and plotting, we can use a ``sphere_proj`` in
exactly the same way as we would a ``line_proj``, but with one key
difference. The binning arguments of ``cut`` no longer refer to
:math:`h,k,l,E`, but to |Q|, :math:`\theta`, :math:`\phi`, :math:`E`.

.. code-block:: matlab

   sp_cut = cut(w, sp_proj, Q, theta, phi, e, ...);

.. warning::

   The form of the arguments to ``cut`` is still the same (see: `Binning

>    arguments`_), however:

Reviewer: ```suggestion
   arguments`_). However:
```

   - |Q| runs from :math:`[0, \infty)` (attempting to use a |Q| with a minimum
     bound less than :math:`0` will result in unexpected behaviour)
   - :math:`\theta` runs between :math:`[0, 180)`
   - :math:`\phi` runs between :math:`[-180, 180)`
   - Attempting to use default binning (``[]``, ``[n]``) will produce
     erroneous results if source object is in a linear projection

>      (``line_proj``).

Reviewer: above 2 lines are about line_proj not sphere_proj

Author: No, they're about taking a cut using default binning arguments from an `sqw` (the source object) which is in a linear projection (`line_proj`)  when your target cut is in a spherical projection.

Author: Having spoken to both Christian and Mo, they think the meaning here is clear.

where:

- ``ez``

  is :math:`e_z` the 3-vector defining the direction of the :math:`z`-axis in
  reciprocal space of the spherical coordinate system.

- ``ex``

  is :math:`e_x` the 3-vector defining the direction of the :math:`x`-axis in
  reciprocal space of the spherical coordinate system.

.. note::

   The :math:`e_y` direction is not explicitly defined as part of the
   ``sphere_proj`` and is always considered to be perpendicular to the
   :math:`e_x`-:math:`e_z` plane.

.. note::

   By default a ``sphere_proj`` will define its principal axes
   :math:`e_z` and :math:`e_x` along the :math:`hkl` directions
   :math:`[1,0,0]` (:math:`e_z`) and :math:`[0,1,0]` (:math:`e_x`)
   respectively.

- ``type``

  Three character string denoting the the projection normalization of each
  dimension, one character for each, e.g. ``'add'``, ``'rrr'``, ``'pdr'``.

  There are 3 possible options for the first (length) component of ``type``,
  which are the same as those for ``line_proj``:

  1. ``'a'``

     Inverse angstroms

  2. ``'r'``

     Reciprocal lattice units (r.l.u.) which normalises so that the maximum of
     :math:`|h|`, :math:`|k|` and :math:`|l|` is unity.

  3. ``'p'``

> 

Reviewer 2: Everything excect `a`  is not currently there and I am not sure what is written here now has physical meaning.  certainly `p` would mention `e_z` if implemented as `u` and `v` are not used in `sphere_proj` and `r` will also mean other thing. Instead of guessing now, I would just remove it.

Reviewer 2: May be `r` would mean normalizing `e_z` by its length in Angstroms and `p` is irrelevant as `e_z` is always directed along specified direction, so currently I see only two reasonable meaning for first value

     Preserve the values of ``u`` and ``v``

  There are 2 possible options for the second and third (angular) components of
  type:

  1. ``'d'``

     Degrees

  2. ``'r'``

     Radians

  For example, if we wanted the **Q**-component to be in r.l.u. and
  the angles in degrees we would have ``type = 'rdd'``.

- ``alatt``

  3-vector of lattice parameters.

- ``angdeg``

  3-vector of lattice angles in degrees.

.. note::

   In general, you should not need to define ``alatt`` or ``angdeg``;
   by default they will be taken from the ``sqw`` object during a
   ``cut``. However, there are cases where a projection object may
   need to be reused elsewhere.

- ``offset``

  3-vector in :math:`(h,k,l)` or 4-vector in :math:`(h,k,l,e)` defining the
  origin of the projection coordinate system.

.. warning::

   Prior to Horace 4.0.0 a similar (mis-named) property ``uoffset`` was
   used. This will be ignored in Horace 4.0.0, so ensure that your scripts have

>    been updated.

Reviewer 2: I have reconciled that so `offset` in Horace-4 and `uoffset` in Horace-3 are mapped to the same property implicityly. This warning should be safely removed

   rescaled in |Q| relative to the magnitude of :math:`e_x` or
   :math:`e_z`.

When it comes to cutting and plotting, we can use a ``sphere_proj`` in
exactly the same way as we would a ``line_proj``, but with one key
difference. The binning arguments of ``cut`` no longer refer to
:math:`h,k,l,E`, but to |Q|, :math:`\theta`, :math:`\phi`, :math:`E`.

.. code-block:: matlab

   sp_cut = cut(w, sp_proj, Q, theta, phi, e, ...);

.. warning::

   The form of the arguments to ``cut`` is still the same (see: `Binning
   arguments`_), however:

   - |Q| runs from :math:`[0, \infty)` (attempting to use a |Q| with a minimum
     bound less than :math:`0` will result in unexpected behaviour)
   - :math:`\theta` runs between :math:`[0, 180)`
   - :math:`\phi` runs between :math:`[-180, 180)`
   - Attempting to use default binning (``[]``, ``[n]``) will produce
     erroneous results if source object is in a linear projection
     (``line_proj``).

``sphere_proj`` 2D and 1D cuts examples:
________________________________________

By integrating over the angular terms of a spherical projection you
can obtain the powder average of a sample (excluding certain powder

> effects if done on a single crystal).

Reviewer 2: Are you sure? I am not aware of any powder effects excluded. I thougt you are building excact powder average. 
The effects should be either clarified or sentence removed.

Author: This comment came I believe out of a discussion with Russell will double check at lunch

Reviewer 3: I think a clearer wording would be:

```
Integrating over the angular terms of a spherical projection of a
**single crystal** dataset will give an **approximation** of a powder
average of the sample. This is because (except for low scattering
angles) the detectors do not cover the full 4*pi solid angular range.
Thus regions without detector coverage will not be sampled by
the angular spherical integration. In contrast for a true powder
sample, there will be crystal grains with the correct orientation to
be detected even by the limited detector coverage.
Integrating over the angular terms for a *powder* sample is
a valid powder averaging.

At low scattering angles (below approximately 30 degrees on LET),
the detectors *do* cover the full angular range so the angular
integration of a single crystal dataset will give a valid powder
average.

These effects are important to bear in mind when modelling the
scattering - e.g. for a single crystal dataset it is best to model it
as a single crystal and then let Horace perform the angular
integration, rather than treating it as powder data.
```

Sorry it's a bit long-winded... perhaps you can make it terser.

      angdeg % => [60 60 90]
      proj = line_proj([1 0 0], [0 1 0]);

   such that ``u`` = :math:`[1,0,0]` and ``v`` = :math:`[0,1,0]`. The
   reciprocal space projection will actually be skewed according to ``angdeg``.

.. _nonortho:

* ``nonorthogonal``

  Whether lattice vectors are allowed to be non-orthogonal

.. note::

  If you don't specify ``nonorthogonal``, or set it to ``false``, you will get
  orthogonal axes defined by ``u`` and ``v`` normal to ``u`` and ``u`` x
  ``v``. Setting ``nonorthogonal`` to ``true`` forces the axes to be exactly the ones
  you define, even if they are not orthogonal in the crystal lattice basis.

  .. warning::

     Any plots produced using a non-orthogonal basis will plot them as though
     the basis vectors are orthogonal, so features may be skewed (see
     `below <#non-orthogonal-axes>`_) .

     The benefit to this is that it makes reading the location of a feature in a
     two-dimensional **Q**-**Q** plot straightforward. This is the main reason for

>      treating non-orthogonal bases this way.

Author: That's what the see: below does.

   :width: 500px
   :alt: Q-phi cut

   MAPS Fe data; Spin-wave scattering intensity the the origin centred
   about the :math:`[0,-1,1]` Bragg peak. A visible gap caused by
   missing detectors is obvious in the :math:`\phi`-axis range
   :math:`[-50^\circ:+50^\circ]`.

>    Inset: Linear projection of the same region, the red lines show the

Reviewer: ```suggestion
   Inset: Linear projection of the same region; the red lines show the
```

   approximate mapping from the linear to spherical projections.

> Integrating over all :math:`\theta` and thus including other detectors

Reviewer: ```suggestion
Integrating over the whole :math:`\theta` range and thus including other detectors
```

> substantially improves statistics, this is done by setting the

Reviewer: ```suggestion
substantially improves statistics; this is done by setting the
```

:math:`\theta` parameter to ``[0, 180]``:

.. code-block:: matlab

    s2 = cut(data_source, sp_proj, [0, 0.1, 2], [0, 180], [-180, 4, 180], [50, 60]);

.. figure:: ../images/spin_w_theta_av.png
   :align: center
   :width: 500px
   :alt: Q-phi cut theta-averages

   MAPS Fe data; Scattering intensity from cut averaged over all :math:`\theta`
   spin-wave with the origin centred at the :math:`[0,-1,1]` Bragg
   peak.

> The 1D cut below, generated by integrating over the :math:`\phi`-axis,

Reviewer: and theta, yes?

Author: Yes and a subsection of energy, but this is a series of operations and the previous plot was already integrated over all theta.

Further Examples
----------------

> To take a cut from an existing ``sqw`` or ``dnd`` object retaining the existing

Reviewer: ```suggestion
To take a cut from an existing ``sqw`` or ``dnd`` object while retaining the existing
```

> projection, omit the ``proj`` argument:

Reviewer: to be pedantic, it's not omitted, it's blanked with the empty object

Author: ```suggestion
projection, provide an empty ``proj`` argument:
```
